gh-118610: Centralize power caching in `_pylong.py`

tim-one opened the pull request:
A new internal function computes, in advance, all and only the powers various other functions will need. While a few percent faster overall (timing roundtrip `int(str(n))` on "large" n), the primary point is to free the other functions from needing to embed their own ad-hoc caching schemes.

Still a work in progress. For now, both the old and new versions of the client functions are present, and new `setold()` and `setnew()` functions allow dynamically switching between them. That's for testing, and all that will eventually be removed.

<!-- gh-issue-number: gh-118610 -->
* Issue: gh-118610
<!-- /gh-issue-number -->

tim-one committed:
Work in progress. Change functions to use a new function
to cache the powers they need.

tim-one committed:
New test that `n == int(str(n))` for large random n.

tim-one committed:
Merge branch 'main' into intconv

tim-one committed:
Merge remote-tracking branch 'upstream/main' into intconv

tim-one committed:
Merge branch 'main' into intconv

tim-one committed:
Fixed the logic so that `**` is used only for the smalles power
needed, and added asserts to ensure that's so.

tim-one committed:
Merge branch 'intconv' of https://github.com/tim-one/cpython into intconv

tim-one committed:
I typed "1009" by mistake instead of "1000". This could cause the
newer int->str function to receive fewer powers than it needed.

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -19,6 +19,82 @@
+        seen.add(w)
+        lo = w >> 1
+        hi = w - lo
+        # only _nned_ lo here; some other path may, or may not,
```
```suggestion
        # only _need_ lo here; some other path may, or may not,
```

tim-one replied:
Thanks, but I already made that change :smile:.

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -83,6 +159,50 @@ def inner(n, w):
+            return D(n)
+        w2 = w >> 1
+        hi = n >> w2
+        lo = n - (hi << w2)
```
It is not related to this issue, but I wonder, would not `lo = n & ((1 << w2)-1)` be faster?

tim-one replied:
I don't know. From previous experience, I know that the relative speeds of CPython's bit-fiddling operations varies a whole lot across platforms. I picked the spelling that requires 2 bigint operations instead of the 3 in your alternative, minimizing the number of trips around the eval loop :wink:.

In this context, though, it doesn't matter - both are `O(n.bit_length())`, insignificant compared to the multiply on the next line.

tim-one replied:
Trying it, your suggestion is significantly faster for very large ints. So I'll adopt it. Thanks! I got into the habit for code using small ints, where minimizing the number of trips around the eval loop can actually matter.

serhiy-storchaka reviewed (commented):
If the goal is simplification of the code, I do not see how is this change an improvement. My implementation of `int_to_decimal_string()` is about 20 lines long (ignoring comments), and the caching of `10**w` adds only 3 lines. `w2pow()` and `w5pow()` add 12 lines each. `compute_powers()` is almost 50 lines long.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into intconv

tim-one committed:
Remove the recursion in compute_powers().

tim-one committed:
Added a comment, and adopted Serhiy's suggestion to speed "lo" extraction.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into intconv

tim-one committed:
Merge branch 'main' into intconv

tim-one committed:
Remove the old functions, and the ability to switch implementations.

That was for running comparative timing tests. & that's over.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into intconv

tim-one committed:
In the new test, allow random changes to the bit length
to be negative too, and check that the result string doesn't
start with a zero.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into intconv

serhiy-storchaka commented on Lib/test/test_int.py:
```diff
@@ -906,6 +906,18 @@ def test_pylong_misbehavior_error_path_from_str(
+        while bits <= 1_000_000:
+            bits += randrange(-100, 101) # break bitlength patterns
+            hibit = 1 << (bits - 1)
+            n = hibit + randrange(hibit)
```
'getrandbits(bits - 1)` can be slightly faster.
```suggestion
            n = hibit | getrandbits(bits - 1)
```

tim-one replied:
The savings would be trivial compared to the two bigint base conversions that follow it, and the enclosing loop doesn't even go around 10 times ... but, ya, I think it's a little _clearer_, so I'll accept it. Thanks!

serhiy-storchaka commented on Lib/test/test_int.py:
```diff
@@ -906,6 +906,18 @@ def test_pylong_misbehavior_error_path_from_str(
                 int(big_value)
 
+    def test_pylong_roundtrip(self):
+        from random import randrange
```
```suggestion
        from random import randrange, getrandbits
```

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -144,38 +197,17 @@ def _str_to_int_inner(s):
+    w5pow = compute_powers(len(s), 5, DIGLIM)
     return inner(0, len(s))
 
-
```
These removals of empty lines are perhaps not intentional.

tim-one replied:
Do you care? I don't.

tim-one replied:
OK - seems pointless to me :wink:, but I put the useless blank line back in.

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -19,6 +19,86 @@
+        else:
+            lo = this >> 1
+            hi = this - lo
+            assert lo in d
```
It is not needed, because `d[lo]` will raise an exception if this condition is false.

tim-one replied:
Yes, but It communicates intent _explicitly_. I want to emphasize it, not leave it to the reader to deduce.

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -33,57 +113,32 @@ def int_to_decimal(n):
+    BITLIM = 200
 
+    # Don't bother caching the "lo" mask in this; the time to compute it is
+    # tiny comopared to the multiply.
```
```suggestion
    # tiny compared to the multiply.
```

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -115,7 +167,9 @@ def inner(n, w):
     # the 52-bit physical address limit in both Intel64 and AMD64.
     w = int(w * 0.3010299956639812 + 1)  # log10(2)
-    pow10_cache = {}
+    pow10 = compute_powers(w, 5, 1000)
```
Now we may want to introduce a named constant for `1000` which is used twice.

tim-one replied:
Good catch! Done.

serhiy-storchaka reviewed (approved):
Looks correct to me.

When you finish with this, please do not forget to edit the commit message when merging the PR.

tim-one committed:
Update Lib/_pylong.py

Spelling error.

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

tim-one committed:
Update Lib/test/test_int.py

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

tim-one committed:
Update Lib/test/test_int.py

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

tim-one committed:
Add named constant for int_to_decimal_string's digit limit.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into intconv

tim-one committed:
Merge branch 'main' into intconv

tim-one committed:
Merge branch 'main' into intconv
